certbot-apache/certbot_apache/augeasparser.py

```diff
@@ -0,0 +1,69 @@
+""" Augeas implementation of the ParserNode interface """
+from certbot_apache import assertions
+from certbot_apache import interfaces
+from certbot_apache import parsernode_util as util
+
+
+class AugeasParserNode(interfaces.ParserNode):
+    """ Augeas implementation of ParserNode interface """
+
+    def __init__(self, **kwargs):
+        ancestor, dirty, filepath, metadata = util.parsernode_kwargs(kwargs)  # pylint: disable=unused-variable
+        super(AugeasParserNode, self).__init__(**kwargs)
+        self.ancestor = ancestor
+        # self.filepath = filepath
+        self.filepath = assertions.PASS
+        self.dirty = dirty
+        self.metadata = metadata
+
+    def save(self, msg): # pragma: no cover
+        pass
+
+
+class AugeasCommentNode(AugeasParserNode):
+    """ Augeas implementation of CommentNode interface """
+
+    def __init__(self, **kwargs):
+        comment, kwargs = util.commentnode_kwargs(kwargs)  # pylint: disable=unused-variable
+        super(AugeasCommentNode, self).__init__(**kwargs)
+        # self.comment = comment
+        self.comment = assertions.PASS
+
+    def __eq__(self, other): # pragma: no cover
+        if isinstance(other, self.__class__):
```

Comment:
So presumably this equality check is written for the `assertEqualComment` above. In which case, we'll want to be able to check equality of comments that are implemented by different implementations of the ParserNode interface. Therefore it's probably ok for them to be equal even if `__class__` isn't equal. Also, it seems a little weird to be checking equality of parent instance variables here, especially when we have that whole thing about `filepath` above. So maybe we could call a `super` equality check, and implement `__eq__` in `AugeasParserNode`.
 
Actually thinking about this, it might make sense to have a `CommentNode` interface as well (though that may be overkill). But either way, if the equality check is going to fail in `assertEqualComment` anyway, what is this check for?

Reply:
This is meant for checking the actual equality of two different instances of same implementation. In some cases we might need to omit comparing some details (first thing that comes to mind is whitespacing within metadata). So I think it should stay as it is.

Reply:
In that case, would you mind explaining what is for?

https://github.com/certbot/certbot/blob/62fd2d5753785247b201560c83b31a502b581c30/certbot-apache/certbot_apache/assertions.py#L33

Reply:
We're expecting the two implementations to be different ones (the other being AugeasParserNode, while other one being ApacheConfigParserNode when it's created).

Reply:
We also want the actual equality assertion kick in only if the `PASS` string is not present in either of the instances.

Reply:
> The current situation where both; primary and secondary are instances of the same implementation is only temporary placeholder.

So is the plan to change this eq method later, once we have `ApacheConfigParserNode` implemented?

Reply:
There shouldn't be a need for the change. The assertion in https://github.com/certbot/certbot/blob/62fd2d5753785247b201560c83b31a502b581c30/certbot-apache/certbot_apache/assertions.py#L33
Is comparing the `comment` property of the `AugeasCommentNode` instances instead of the complete instances for this exact reason.